File: scripts/pi_anonymizer.R
#' PI Name Generator and Anonymizer
#'
#' Provides functions in R to generate random synthetic PI names or anonymized
#' tokens, and to map real PI names to confidential tokens stored in secrets/pi_mapping.json.

#' Generate Random PI Names or Tokens
#'
#' @param n Integer, number of names to generate (default 1).
#' @param format Character, either "synthetic" (realistic surnames) or "token" (PI_i, PI_j...).
#' @param prefix Character, prefix when format is "token" (default "PI_").
#' @param seed Optional integer, random seed for reproducibility.
#' @return Character vector of length n (or scalar if n = 1).
#' @export
generate_pi_names <- function(n = 1,
                              format = c("synthetic", "token"),
                              prefix = "PI_",
                              seed = NULL) {
  if (!is.null(seed)) set.seed(seed)
  format <- match.arg(format)
  
  synthetic_names <- c(
    "Khan", "Armstrong", "Zhao", "Smith", "Patel", "Garcia", "Kim",
    "Chen", "Williams", "Johnson", "Brown", "Miller", "Davis",
    "Wilson", "Martinez", "Anderson", "Taylor", "Thomas", "Hernandez",
    "Moore", "Martin", "Jackson", "Thompson", "White", "Lopez",
    "Lee", "Gonzalez", "Harris", "Clark", "Lewis", "Robinson", "Walker"
  )
  
  if (format == "token") {
    # Generate cryptographic hash tokens: PI_... with 256 bits of CSPRNG entropy
    tokens <- vapply(seq_len(n), function(i) {
      h <- as.character(openssl::sha256(openssl::rand_bytes(32)))
      paste0(prefix, substr(h, 1, 16))
    }, character(1))
    return(if (n == 1) tokens[1] else tokens)
  }
  
  sampled <- sample(synthetic_names, size = n, replace = TRUE)
  if (n == 1) sampled[1] else sampled
}

#' Default secrets path (user data directory, not the repo)
default_secrets_path <- function() {
  if (requireNamespace("rappdirs", quietly = TRUE)) {
    dir <- rappdirs::user_data_dir("TempleCBE")
  } else {
    dir <- file.path(Sys.getenv("HOME", unset = tempdir()), ".TempleCBE")
  }
  if (exists("validate_secrets_dir", mode = "function")) {
    validate_secrets_dir(dir)
  } else {
    if (!dir.exists(dir)) dir.create(dir, recursive = TRUE, showWarnings = FALSE)
  }
  file.path(dir, "pi_mapping.json")
}

# helper: find repo root by locating .git
.find_repo_root <- function(path = ".") {
  p <- tryCatch(normalizePath(path, winslash = "/", mustWork = FALSE), error = function(e) NULL)
  if (is.null(p)) return(NULL)
  repeat {
    if (dir.exists(file.path(p, ".git"))) return(p)
    parent <- dirname(p)
    if (identical(parent, p)) return(NULL)
    p <- parent
  }
}

#' Retrieve or Map Anonymized Token for a PI Name
#'
#' @param name Real PI name to mask.
#' @param secrets_path Path to confidential mapping file (default "secrets/pi_mapping.json").
#' @return Anonymized string token.
#' @export
anonymize_pi <- function(name, secrets_path = NULL) {
  # If no path provided, use a user-scoped defaults directory (not the repo)
  if (is.null(secrets_path)) {
    secrets_path <- default_secrets_path()
  } else {
    parent_dir <- dirname(secrets_path)
    if (exists("validate_secrets_dir", mode = "function")) {
      validate_secrets_dir(parent_dir)
    }
  }
  secrets_path <- tryCatch(normalizePath(secrets_path, winslash = "/", mustWork = FALSE), error = function(e) secrets_path)

  # Refuse to write into the repository tree to avoid accidentally committing PHI
  repo_root <- .find_repo_root(".")
  if (!is.null(repo_root)) {
    repo_root_norm <- normalizePath(repo_root, winslash = "/", mustWork = FALSE)
    if (startsWith(secrets_path, repo_root_norm)) {
      stop("Refusing to write mapping into repository path (", secrets_path, "). Pass an explicit secrets_path outside the repository to override.")
    }
  }

  write_atomic <- function(content, target) {
    if (exists("atomic_write_file", mode = "function")) {
      atomic_write_file(content, target)
    } else {
      tmp <- tempfile(pattern = ".tmp_", tmpdir = dirname(target))
      writeLines(content, con = tmp)
      file.rename(tmp, target)
    }
  }

  lock_dir <- paste0(secrets_path, ".lock")
  acquire_lock <- function(lpath, timeout = 10) {
    t0 <- Sys.time()
    while (!dir.create(lpath, showWarnings = FALSE)) {
      if (as.numeric(difftime(Sys.time(), t0, units = "secs")) > timeout) {
        stop("Failed to acquire advisory lock on ", lpath)
      }
      Sys.sleep(0.05)
    }
  }

  acquire_lock(lock_dir)
  on.exit(if (dir.exists(lock_dir)) unlink(lock_dir, recursive = TRUE, force = TRUE), add = TRUE)

  # Ensure file exists (create an empty mapping if needed)
  if (!file.exists(secrets_path)) {
    mapping_data <- list(mappings = list())
    if (requireNamespace("jsonlite", quietly = TRUE)) {
      write_atomic(jsonlite::toJSON(mapping_data, pretty = TRUE, auto_unbox = TRUE), secrets_path)
    } else {
      write_atomic('{"mappings": {}}', secrets_path)
    }
  }
  
  if (requireNamespace("jsonlite", quietly = TRUE)) {
    mapping_data <- jsonlite::fromJSON(secrets_path)
    mapping <- if (is.list(mapping_data$mappings)) mapping_data$mappings else list()
    
    if (name %in% names(mapping)) {
      return(mapping[[name]])
    }
    
    # Auto-assign cryptographic hash token with 16 hex chars
    h <- as.character(openssl::sha256(openssl::rand_bytes(32)))
    token <- paste0("PI_", substr(h, 1, 16))
    existing <- as.character(unlist(mapping, use.names = FALSE))
    counter <- 1L
    while (token %in% existing) {
      counter <- counter + 1L
      token <- paste0("PI_", substr(as.character(openssl::sha256(paste0(name, "_", counter))), 1, 16))
    }
    mapping[[name]] <- token
    mapping_data$mappings <- mapping
    write_atomic(jsonlite::toJSON(mapping_data, pretty = TRUE, auto_unbox = TRUE), secrets_path)
    return(token)
  } else {
    lines <- readLines(secrets_path, warn = FALSE)
    match_line <- grep(paste0('"', name, '"\\s*:'), lines, value = TRUE)
    if (length(match_line) > 0) {
      val <- sub('.*:\\s*"([^"]+)".*', '\\1', match_line[1])
      return(val)
    }
    h <- as.character(openssl::sha256(openssl::rand_bytes(32)))
    return(paste0("PI_", substr(h, 1, 16)))
  }
}

File: scripts/pi_anonymizer.py
"""
PI Name Generator and Anonymizer

Provides utilities to generate random synthetic PI names or anonymized tokens,
and to map real PI names to confidential tokens stored in a secure, user-scoped
location outside the repository tree.
"""

import hmac
import hashlib
import json
import os
import random
import secrets
import tempfile
from pathlib import Path
from typing import List, Union, Optional, Sequence

DEFAULT_SYNTHETIC_NAMES = [
    "Khan", "Armstrong", "Zhao", "Smith", "Patel", "Garcia", "Kim", 
    "Chen", "Williams", "Johnson", "Brown", "Miller", "Davis", 
    "Wilson", "Martinez", "Anderson", "Taylor", "Thomas", "Hernandez", 
    "Moore", "Martin", "Jackson", "Thompson", "White", "Lopez", 
    "Lee", "Gonzalez", "Harris", "Clark", "Lewis", "Robinson", "Walker"
]


def find_repo_root(start: Optional[Path] = None) -> Optional[Path]:
    """Find git repository root by traversing upward from start directory."""
    cur = (start or Path.cwd()).resolve()
    for parent in [cur, *cur.parents]:
        if (parent / ".git").exists():
            return parent
    return None


def default_secrets_path() -> Path:
    """Return default user-scoped path for confidential PI mappings outside repo."""
    env_path = os.environ.get("TEMPLECBE_SECRETS_PATH")
    if env_path:
        return Path(env_path).resolve()
    return (Path.home() / ".TempleCBE" / "pi_mapping.json").resolve()


def validate_secrets_path(path: Path) -> Path:
    """Ensure secrets path is resolved and does not reside within the repository tree."""
    resolved = path.resolve()
    repo_root = find_repo_root(resolved.parent)
    if repo_root:
        try:
            resolved.relative_to(repo_root)
            raise ValueError(
                f"Refusing to store confidential mappings inside the repository tree ({resolved}). "
                "Secrets must reside in a user-scoped directory outside the repository (e.g. ~/.TempleCBE/pi_mapping.json)."
            )
        except ValueError as err:
            # If not relative to repo_root, relative_to raises ValueError which is what we want
            if "Refusing to store" in str(err):
                raise err
    return resolved


def generate_pi_names(
    n: int = 1, 
    format: str = "synthetic", 
    prefix: str = "PI_", 
    n_chars: int = 16,
    seed: Optional[int] = None,
    exclude: Optional[Sequence[str]] = None
) -> Union[str, List[str]]:
    """
    Generate random PI names or anonymized tokens.
    
    :param n: Number of names to generate (default 1).
    :param format: 'synthetic' (realistic surnames) or 'token' (cryptographic hash).
    :param prefix: Prefix when format='token'.
    :param n_chars: Number of hex characters from the hash digest (default 16).
    :param seed: Optional random seed for reproducible sampling.
    :param exclude: Optional collection of surnames (case-insensitive) to omit.
    :return: A single name string if n == 1, else a list of strings.
    """
    if seed is not None:
        random.seed(seed)
    
    exclude_set = {e.strip().lower() for e in (exclude or [])}

    if format == "token":
        names = [f"{prefix}{secrets.token_hex(max(1, (n_chars + 1) // 2))[:n_chars]}" for _ in range(n)]
    else:
        pool = [name for name in DEFAULT_SYNTHETIC_NAMES if name.lower() not in exclude_set]
        if not pool:
            pool = [f"Investigator_{i}" for i in range(1, n + 1)]
        names = random.choices(pool, k=n)
        
    return names[0] if n == 1 else names


def anonymize_pi(
    name: str, 
    secrets_path: Optional[Union[str, Path]] = None,
    key: Optional[str] = None,
    n_chars: int = 16,
    prefix: str = "PI_",
    auto_assign: bool = True
) -> str:
    """
    Retrieve or create an anonymized pseudonym token for a real PI name.
    
    Uses an external, confidential mapping JSON file stored outside the repository.
    Tokens are generated via HMAC-SHA256 (or keyed SHA256) matching TempleCBE's R engine.
    
    :param name: Real PI surname or identifier.
    :param secrets_path: Path to confidential mapping file. Defaults to ~/.TempleCBE/pi_mapping.json.
    :param key: Optional HMAC secret key / pepper. Defaults to TEMPLECBE_SECRET_KEY env var.
    :param n_chars: Length of hex digest to retain (default 16).
    :param prefix: Token prefix (default "PI_").
    :param auto_assign: If True, automatically creates and atomically saves a new token.
    :return: Anonymized token (e.g. 'PI_37b51069...').
    """
    path = validate_secrets_path(Path(secrets_path) if secrets_path else default_secrets_path())
    effective_key = key if key is not None else os.environ.get("TEMPLECBE_SECRET_KEY", "temple_cbe_default_salt")

    data = {}
    mapping = {}
    
    if path.exists():
        try:
            with open(path, "r", encoding="utf-8") as f:
                data = json.load(f)
                mapping = data.get("mappings", {})
        except Exception:
            mapping = {}
            
    if name in mapping:
        return mapping[name]
        
    if auto_assign:
        # Keyed HMAC-SHA256 matching R implementation
        h = hmac.new(effective_key.encode("utf-8"), name.encode("utf-8"), hashlib.sha256).hexdigest()
        token = f"{prefix}{h[:n_chars]}"
        mapping[name] = token
        data["mappings"] = mapping
        
        path.parent.mkdir(parents=True, exist_ok=True)
        # Atomic write
        temp_dir = path.parent
        with tempfile.NamedTemporaryFile("w", dir=temp_dir, delete=False, encoding="utf-8") as tf:
            json.dump(data, tf, indent=2)
            temp_name = tf.name
        Path(temp_name).replace(path)
        return token
        
    return name


if __name__ == "__main__":
    print("=== Synthetic PI Names ===")
    print("Single random PI:", generate_pi_names(1))
    print("5 random PIs:", generate_pi_names(5, exclude=["Armstrong"]))
    print("\n=== Anonymized Tokens ===")
    print("Tokens (16-char):", generate_pi_names(4, format="token", n_chars=16))
    print("\n=== Secrets-Backed Lookup (stored outside repo) ===")
    print("Default secrets file:", default_secrets_path())
    for pi in ["Franklin", "Taylor", "NewInvestigator"]:
        print(f"{pi} <-> {anonymize_pi(pi)}")
